### PR TITLE
[CMSP-403] Main terminus command(s)

### PR DESCRIPTION
**Pull request by jazzsequence**
Adds the `addons-install` command and its `install` alias and subcommands.

`install:list` returns a (hard-coded) list of available jobs
`install:run <site>.<env> <job>` runs the specified job on the given environment. Omitting `.<env>` assumes `.dev`
`install` or `install:help` displays a help message with usage information

### CODE REVIEW

File: src/Helpers/UtilityFunctions.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    /**
+     * Given a $site_env which might be in site_id.dev or just site_id format, return an array with the site_id and env.
+     *
+     * The default environment is dev.
+     *
+     * @param string $site_env
+     *
+     * @return array
+     */
+    public static function decypherSiteInfo(string $site_env) : array
```
Comment (pwtyler) [inline src/Helpers/UtilityFunctions.php:22]: Is this not already something in core terminus lib? or is defaulting to dev the fun edge case?

File: tests/unit/UtilityFunctionsTest.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    public function testBaseCommand($expected)
+    {
+        $this->assertTrue(str_contains(UtilityFunctions::usage(), $expected));
+    }
+
+    /**
+     * Data provider for testBaseCommand.
+     *
+     * @return array
+     */
```
Comment (pwtyler) [inline tests/unit/UtilityFunctionsTest.php:23]: 🙌🏻  table driven tests 🙌🏻

**Review by pwtyler — CHANGES_REQUESTED** [review]

File: src/Helpers/UtilityFunctions.php (hunk 1 above)
Comment (jazzsequence) [reply src/Helpers/UtilityFunctions.php:22]: `getSite` from `SiteAwareInterface` expects us to know the `$site_id` already, and `getEnvironments` expects us to know the `$site_env`. When I saw these being invoked elsewhere, it looked to me like this stuff still had to be identified somehow, so I think this helper function is still useful. See https://github.com/pantheon-systems/terminus-secrets-manager-plugin/blob/main/src/Commands/SiteSetCommand.php#L59-L61

File: .bin/cleanup.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+set +ex
```
Comment (pwtyler) [inline .bin/cleanup.sh:2]: clarifying, you wish to _disable_ errexit and xtrace, or did you intend set -ex?

I don't usually set xtrace, but don't see harm in more debugging.

File: .bin/cleanup.sh [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+SITE_ENV="${TERMINUS_SITE}.ci-${BUILD_NUM}"
+# Nuke any lingering multidev environments from orbit.
+for ENV in $(terminus multidev:list --field=Name --format=list $TERMINUS_SITE); do
+  if [[ $ENV == ci-* ]]; then
```
Comment (pwtyler) [inline .bin/cleanup.sh:6]: 🔢 shellcheck would complain about unquoted variables. This comment is not specifically about the quoting, but that it's up to you if you want to apply shellcheck to .bin. I'm generally in favor it.

File: .bin/set-up-globals.sh [hunk 5]
```diff
@@ -1,12 +1,30 @@
 #!/bin/bash
 set -ex
 
-
 TERMINUS_PLUGINS_DIR=.. terminus list -n remote
 
 set +ex
```
Comment (pwtyler) [inline .bin/set-up-globals.sh:6]: once again, unless there's good reason for it, in favor of at least set -e for the whole script and skipping -x.

File: .bin/set-up-globals.sh [hunk 6]
```diff
@@ -1,12 +1,30 @@
...
+	chmod 600 ~/.ssh/config
+fi
+
+echo "Editing the ~/.ssh/config file"
+# Set StrictHostKeyChecking to no in ~/.ssh/config
+echo "StrictHostKeyChecking no" >> ~/.ssh/config
+# Set LogLevel to ERROR in ~/.ssh/config
+echo "LogLevel ERROR" >> ~/.ssh/config
+# Set UserKnownHostsFile to /dev/null in ~/.ssh/config
+echo "UserKnownHostsFile /dev/null" >> ~/.ssh/config
```
Comment (pwtyler) [inline .bin/set-up-globals.sh:30]:
```suggestion
echo "Editing the ~/.ssh/config file"
echo "StrictHostKeyChecking no" >> ~/.ssh/config
echo "LogLevel ERROR" >> ~/.ssh/config
echo "UserKnownHostsFile /dev/null" >> ~/.ssh/config
```

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -15,34 +15,38 @@ jobs:
...
       - name: Checkout
         uses: actions/checkout@v2
       - name: Set up environment
         run: ./.bin/set-up-globals.sh
-      - name: Validate Terminus exists
-        run: terminus --version
       - name: Install plugin
         run: terminus self:plugin:install ${GITHUB_WORKSPACE}
+      - name: Install SSH key
+        uses: webfactory/ssh-agent@v0.7.0
```
Comment (pwtyler) [inline .github/workflows/test.yml:41]: Something I read about today and can't decide if I want to go down this route— pin this to a specific commit hash, as the tag could be compromised and tweet out your secrets.

File: .bin/cleanup.sh [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+	terminus multidev:delete --delete-branch --yes ${TERMINUS_SITE}.${ENV}
+  fi
+  # Also dump any localtests environments.
+  if [[ $ENV == localtests ]]; then
+	terminus multidev:delete --delete-branch --yes ${TERMINUS_SITE}.localtests
+  fi
```
Comment (pwtyler) [inline .bin/cleanup.sh:12]:
```suggestion
  if [[ $ENV == ci-* || $ENV == localtests ]]; then
	terminus multidev:delete --delete-branch --yes ${TERMINUS_SITE}.${ENV}
  fi
```

File: README.md [hunk 9]
```diff
@@ -16,39 +16,34 @@ These commands require no configuration
...
 
-* `--with-db`: If included, the `scaffold_extension` job will be run with a database connection.
-
+* `--skip-db`: If included, the job will be run without a database connection.
+-->
 ## Usage
-* `terminus scaffold-extension:list <site_id>.<env>`
-* `terminus scaffold-extension:run <site_id>.<env> <job> [--with-db]`
+* `terminus addons-install:list`
+* `terminus addons-install:run <site_id>.<env> <job> [--with-db]`
```
Comment (pwtyler) [inline README.md:40]:
```suggestion
* `terminus addons-install:run <site_id>.<env> <job> [--skip-db]`
```

File: .bin/cleanup.sh [hunk 10]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+set +ex
+SITE_ENV="${TERMINUS_SITE}.ci-${BUILD_NUM}"
+# Nuke any lingering multidev environments from orbit.
```
Comment (pwtyler) [inline .bin/cleanup.sh:4]: This means multiple rounds of testing will get in each others way, or kill a multidev while another branch/build/whatever is in-progress. Alternative will be to cleanup only _this build_ and any build older than X days, and separately have a script  that deletes the oldest multidev at the _top_ of the CI run. I did this for a Drupal module somewhere and we could dig that up. Probably a pretty good custom action tbh.

File: tests/functional/addons-install-test.bats [hunk 11]
```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env bats
+
+# Check if TERMINUS_SITE is defined, otherwise define it as terminus-addons-installer-plugin.
+if [ -z "$TERMINUS_SITE" ]; then
+  # Looks like this is a local run. We'll create the localtests multidev ro test this.
+  TERMINUS_SITE=terminus-addons-installer-plugin
+
+  # Check if the localtests environment exists already, otherwise create it.
```
Comment (pwtyler) [inline tests/functional/addons-install-test.bats:8]: We good with "localtests" being a single env so no more than one eng can test locally at a time? what about checking an env variable (but could still default to localtests)

File: tests/functional/addons-install-test.bats [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+  run terminus addons-install:run ${SITE_ENV} bar
+  [[ $output == *"The bar job does not exist"* ]]
+  [ "$status" -eq 1 ]
+
+  run terminus install:run ${SITE_ENV} bar
+  [[ $output == *"The bar job does not exist"* ]]
+  [ "$status" -eq 1 ]
+}
+
+@test "test failure state if command is run with uncommitted filesystem changes" {
```
Comment (pwtyler) [inline tests/functional/addons-install-test.bats:75]: does bats support using +x instead of all these echos?

File: .bin/cleanup.sh (hunk 10 above)
Comment (pwtyler) [reply .bin/cleanup.sh:4]: https://github.com/pantheon-systems/pantheon_advanced_page_cache/pull/51/files#diff-c08821ed2397f2d1cb87ba9b78650da7f45812c1237a699af51be8776a8b8d55

**Comment by pwtyler** [comment]
Something I just thought of— if this is run on `test` or `live`, I'm not sure where the workflow will breakdown at the moment (likely flipping sftp mode)— but it would be good to have a check here that fast-fails if tried to run against test or live.

File: .bin/cleanup.sh (hunk 10 above)
Comment (pwtyler) [reply .bin/cleanup.sh:4]: "any build older than X days" should be a cron task, not part of any on-push task or anything like that, to keep time of test iteration down.

File: tests/functional/addons-install-test.bats (hunk 12 above)
Comment (jazzsequence) [reply tests/functional/addons-install-test.bats:75]: Not sure, but it's worth noting that the echoes only actually display if the test is failing. An "echo" inside a test does not render otherwise.

File: .bin/cleanup.sh (hunk 3 above)
Comment (jazzsequence) [reply .bin/cleanup.sh:2]: Probably a typo for -ex

File: .bin/cleanup.sh (hunk 10 above)
Comment (jazzsequence) [reply .bin/cleanup.sh:4]: we can pull this out since the cleanup step is always being run. there are a lot of other places where we don't actually clean up after failed tests.

File: tests/functional/addons-install-test.bats (hunk 12 above)
Comment (jazzsequence) [reply tests/functional/addons-install-test.bats:80]: I've removed them anyway as I didn't need them besides for debugging.

File: .github/workflows/test.yml (hunk 7 above)
Comment (jazzsequence) [reply .github/workflows/test.yml:37]: We could also do `webfactory/ssh-agent@master` but that seems sketchy. I would worry that pinning to a hash would be harder to maintain, but I guess not any harder than updating the release number (which, by the by, probably could/should be `0.8.0`, too...)

**Comment by jazzsequence** [comment]
> Something I just thought of— if this is run on `test` or `live`, I'm not sure where the workflow will breakdown at the moment (likely git push)— but it would be good to have a check here that fast-fails if tried to run against test or live.

Let's put that into another PR and get this one out. I can add that later, but this PR is already too big and blocking pushing to Packagist.

**Comment by jazzsequence** [comment]
NM ☝️  doing it anyway.

File: .bin/set-up-globals.sh [hunk 13]
```diff
@@ -1,12 +1,35 @@
...
 
 TERMINUS_PLUGINS_DIR=.. terminus list -n remote
 
-set +ex
-# TODO: $TERMINUS_SITE is not set yet. We still need to set up a CI fixture for running tests against.
-echo "Test site is $TERMINUS_SITE"
 echo "Logging in with a machine token:"
 terminus auth:login -n --machine-token="$TERMINUS_TOKEN"
 terminus whoami
+terminus multidev:create "$TERMINUS_SITE".dev ci-"$BUILD_NUM"
```
Comment (rwagner00) [inline .bin/set-up-globals.sh:9]: This many be a bit bike-sheddy, but does building multidev environments belong in the "set up globals" script?

File: .github/workflows/test.yml (hunk 7 above)
Comment (pwtyler) [reply .github/workflows/test.yml:37]: Yeah it boils down to "do you/we trust this action with my secrets"? a branch or tag could be changed to nefarious code, but a specific git hash would not.

File: .bin/set-up-globals.sh (hunk 13 above)
Comment (jazzsequence) [reply .bin/set-up-globals.sh:9]: ¯\_(ツ)_/¯ the script filename isn't super representative of it's current purpose. I believe it was a file that was created as part of the `terminus plugin:create` command that just got repurposed. It could be re-named to just `setup.sh`

**Review by rwagner00 — APPROVED** [review]

**Review by pwtyler — DISMISSED** [review]

File: .github/workflows/test.yml (hunk 7 above)
Comment (jazzsequence) [reply .github/workflows/test.yml:37]: I'm just paranoid enough to try it...
